mqtt/src/main/java/com/hazelcast/jet/contrib/mqtt/MqttBaseBuilder.java

```diff
@@ -0,0 +1,207 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.contrib.mqtt;
+
+import com.hazelcast.function.SupplierEx;
+import com.hazelcast.internal.util.UuidUtil;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Base class for Mqtt source and sink builder.
+ * See {@link MqttSourceBuilder} and {@link MqttSinkBuilder}.
+ *
+ * @since 4.3
+ */
+abstract class MqttBaseBuilder<T> {
+
+    private static final String LOCAL_BROKER = "tcp://localhost:1883";
+
+    String broker = LOCAL_BROKER;
+    String clientId = UuidUtil.newUnsecureUuidString();
+    String topic;
+    boolean autoReconnect;
+    boolean keepSession;
+    String username;
+    char[] password;
+    SupplierEx<MqttConnectOptions> connectOptionsFn;
+
+    /**
+     * Set the address of the Mqtt broker.
+     * <p>
+     * For example, to connect to the local broker with a non-default port:
+     * <pre>{@code
+     * builder.broker("tcp://localhost:1993")
+     * }</pre>
+     * <p>
+     * Default value is {@link #LOCAL_BROKER}.
+     *
+     * @param broker the address of the server to connect to, specified as a URI.
+     */
+    @Nonnull
+    public MqttBaseBuilder<T> broker(@Nonnull String broker) {
+        this.broker = broker;
+        return this;
+    }
+
+    /**
+     * Set the id of the client. The connector appends global processor index
+     * to the given id to make each created client unique.
+     * <p>
+     * For example, each created client will have: `the-client-0`,
+     * `the-client-1`, `the-client-2`...
+     * <pre>{@code
+     * builder.clientId("the-client")
+     * }</pre>
+     * <p>
+     * Default value is a randomly generated uuid.
+     *
+     * @param clientId the unique id of the client.
+     */
+    @Nonnull
+    public MqttBaseBuilder<T> clientId(@Nonnull String clientId) {
+        this.clientId = clientId;
+        return this;
+    }
+
+    /**
+     * Set the topic which the connector will use.
+     * See {@link MqttSourceBuilder#topic} or {@link MqttSinkBuilder#topic}.
+     */
+    @Nonnull
+    public MqttBaseBuilder<T> topic(@Nonnull String topic) {
+        this.topic = topic;
+        return this;
+    }
+
+    /**
+     * Set that the client automatically attempt to reconnect to the broker
+     * if the connection is lost.
+     * <p>
+     * For example:
+     * <pre>{@code
+     * builder.autoReconnect()
+     * }</pre>
+     * <p>
+     * By default the client will not attempt to reconnect.
+     * If {@link #connectOptionsFn(SupplierEx)} is set, this parameter is
+     * ignored.
+     * <p>
+     * See {@link MqttConnectOptions#setAutomaticReconnect(boolean)}.
+     */
+    @Nonnull
+    public MqttBaseBuilder<T> autoReconnect() {
+        this.autoReconnect = true;
```

Comment:
Automatic reconnection is only one of the reconnect behaviour options offered by `org.eclipse.paho.client.mqttv3.MqttConnectOptions`. Maybe we could cover more and expose our version of the setting as a `com.hazelcast.jet.retry.RetryStrategy`, even if with some limitations?

Reply:
for a more custom configuration user can always use `connectOptionsFn` and provide their own `MqttConnectOptions` instance. In that case other connection related options such as `autoReconnect`, `keepSession` and `auth` will be ignored.

Reply:
I still think that we should use `RetryStrategy` in our own facace, the source/sink builders.